### DISCUSSION TITLE
[RFC] Boundaries

### DISCUSSION THREAD

**Discussion by anthonyshew** [vercel/turborepo#9435] [RFC]
We’d like to introduce a Boundaries feature in Turborepo.

The JavaScript ecosystem has grown into having great support for monorepos, letting developers scale JavaScript and TypeScript repositories like never before. However, as a repository progresses to greater scales, it becomes important to enforce boundaries within the repository to maintain order and predictability in the codebase.

# Try it out (Update 2/13/2025)

Turborepo 2.4.2 has our first experimental implementation of this feature! It includes the default boundaries and custom tags that you can create.

[See the documentation on how to try it out](https://turbo.build/repo/docs/reference/boundaries) and please continue leaving your feedback on this RFC.

## Problems we want to solve

Without clear boundaries in a monorepo, it’s too easy to use code from internal packages in undesirable ways:

- Because code is all on the same file system, you can easily `../../` out of the root of a package. This breaks out of the model of JavaScript monorepos, since code now depends on other code without respecting the package graph. This can result in unexpected caching behavior, since Turborepo can’t account for code outside of the package graph in hashing.
- You can use any package in any other package. This can create entanglement in the dependency graph, creating disorder and confusion that harms developer experience.
- You can easily install an application into another application. This is likely to create a litany of issues when it comes to building and deploying your applications.
- Packages can be used in contexts they aren’t meant for. For example, backend code can be authored much differently than frontend code, without concern for bundle sizes. This can cause the backend-minded package to bloat frontend bundles. At scale, this can be disastrous for web application performance.
- Domain concerns mean that you could have a set of packages that are only meant for certain domain contexts. For example, you may have an `@repo/shared-marketing` package that can accidentally be used in an application that is outside of the marketing domain.

## Past proposals and updates
Below are proposals and iterations we've designed according to your feedback. Please see above for current state of the feature.

<details>
<summary>First spec proposal (description only, not implemented)</summary>
# Proposed Solution

The `turbo boundaries` command will:

- Exit with a 0 exit code to indicate a successful run when no Boundaries have been breached
- Exit with a 1 exit code to indicate failure when a Boundary has been breached, printing output to stderr to indicate where there are issues.

```
turbo boundaries

2 Boundary issues found:
- `@repo/be-utils` is used in `@repo/ui`, but is not
  listed in `dependencies` in `@repo/ui`'s package.json.
  Location: ./packages/ui/src/button.tsx - Line 2, Column 9

- The path `../../.eslint-config.js` leads outside of the package's directory.
  This can lead to caching issues.
  Location: ./packages/ui/.eslint-config.js - Line 1, Column 4

- `@repo/ui` denies packages with the tag `util` to be used. `@repo/utils` has this tag.
  Location: ./packages/ui/package.json - Line 32, Column 8
```

Additionally, any non-CI invocation of `turbo run` or `turbo watch` will include a synthetic "Boundaries checking" task once the Boundaries feature has been enabled. This ensures that we can notify you as soon as possible after writing code that breaches a Boundary.

## Getting started with zero-configuration

Add `boundaries: true` to your root `turbo.json` to start using the recommended checks provided by Turborepo. (This is not required to use [Boundary Tags](#incrementally-adopting-boundary-tags).)

```jsx
// ./turbo.json
{
  "boundaries": true
}
```

This will enforce our recommended set of Boundaries for your Workspace:

- You will no longer be able to path outside of a package’s directory. Assuming `apps/web` is a package’s directory, the code below would result in an error.

```jsx
// apps/web/foo.ts
import {} from "../../../"
```

- Importing code from an internal package that is not installed in the package. Assuming that `@repo/utils` is not listed in `apps/web/package.json#dependencies`, the code below would result in an error. This helps with some package managers that allow dependencies to be available where they aren't installed.

```jsx
import {} from "@repo/utils"
```

Once added, you may find existing issues that you’re not ready to address quite yet. See the [Allowlisting existing code section](#creating-exceptions-for-existing-code) for more on how to incrementally adopt Boundaries.

## Incrementally adopting Boundary Tags

Additionally, you can create Tags to mark domains, concerns, contexts, and other useful sets that you want to make custom Boundaries for.

In a [Package Configuration](https://turbo.build/repo/docs/reference/package-configurations), add a `boundaries` key.

```json
// ./packages/ui/turbo.json
{
  "extends": ["//"],
  "boundaries": {
    "tags": ["library", "ui", "frontend", "stable"]
    // Mutually exclusive. You may only use one of these.
    "allowDependencyOn": ["library", "frontend"],
    "denyDependencyOn": ["backend", "marketing", "alpha", "beta"],
    // Mutually exclusive. You may only use one of these.
    "allowDependencyFrom": ["library", "frontend"],
    "denyDependencyFrom": ["backend"]
  }
}
```

The schema for `boundaries` above represents:

- `tags` : The tags for the package. These will be used for comparisons against the Boundaries of other packages.
- `allowDependencyOn`: List of tags that internal dependencies of the package *must* include *at least one of* in its own tag list, including indirect dependencies. Any package that does not have any of these tags will be considered an error. Cannot be used with `denyDependencyOn`.
- `denyDependencyOn`: List of tags that internal dependencies of the package *cannot include any of* in its own tag list, including indirect dependencies. Any package with any of these tags will be considered an error. Cannot be used with `allowDependencyOn`.
- `allowDependencyFrom`: List of tags that internal dependents *must include at least one of* in their own tag list, including indirect dependencies. Any package that tries to use this package but is not tagged with any members of the `allowDependencyFrom` array will be considered an error. Cannot be used with `denyDependencyOn`.
- `denyDependencyFrom`: List of tags that internal dependents *most not include any of* in their own tag list, including indirect dependencies. Any package that tries to use this package that is tagged with any of the members of `denyDependencyFrom` will be considered an error. Connot be used with `allowDependencyFrom`.

## Creating exceptions for existing code

Introducing Boundaries to a large codebase can be difficult when you want to enforce a new Boundary where there wasn’t one before, so we want to make sure that Boundaries are incrementally adoptable.

After creating a Boundary, you can quickly create an exception list for existing violations by adding an `--exclude-errors` flag.

```
turbo boundaries --exclude-errors
Allowlisted 38 errors.
```

This will create an allowlist file in packages that have violations. The exclusion list would look something like:

```json
// packages/ui/turbo.json
{
  "extends": ["//"],
  "boundaries": {
    "exclusions": [
      {
        "type": "Path breakout",
        "file": "./packages/ui/src/button.tsx",
        "location": "@repo/ui-@repo/utils",
        "reason": "The path `../../../` terminates outside of the package."
      },
      {
        "type": "Missing dependency",
        "file": "./packages/ui/src/card.tsx",
        "location": "@repo/ui-@repo/utils",
        "reason": "`@repo/utils` is used but is cannot be found in package.json"
      },
      {
        "type": "Tag violation",
        "file": "./packages/ui/package.json",
        "location": "@repo/ui-@repo/utils",
        "reason": "`@repo/ui` denies packages with the tag `util` to be used. `@repo/utils` has this tag."
      }
    ]
  }
}
```

Exclusions will also be inspected whenever Boundaries checking occurs (`turbo boundaries`, `turbo run`, `turbo watch`) and errors will be thrown for exclusions that aren’t being used. `turbo boundaries --exclude-errors` will automatically fix the erroneous exclusions.

# Open questions

- We anticipate that the community will want this feature most for Multi-package Workspaces. However, it’s possible this feature could make sense in single-package Workspaces at a module level, as well. We would love to hear use cases where you believe Boundaries could help your single-package Workspace.
- Automatically running the Boundaries check seems like it should not happen in CI, since you'd get the same failure many times. It seems we should teach developers to run `turbo boundaries` as a CI check, and allow all other `turbo run` invocations to proceed without checking Boundaries so that developers can get feedback from their other tasks sooner.
</details>

**Comment by kasper573** [comment]
This is really exciting! I've been using dependency-cruiser and eslint-plugin-monorepo-cop to painstakingly emulate something like this. Having something like this directly in turbo sounds awesome. Nice work!

i agree that boundary checks should be opt in for the ci.

**Comment by reconsumeralization** [comment]
I want in.

**Comment by franky47** [comment]
I could see the value of this kind of linting step, and it being integrated with the rest of package relationship configs (I had missed the fact we could have multiple `turbo.json`, which is nice too) makes sense.

I'd be interested to give this a try in [nuqs](https://github.com/47ng/nuqs), particularly to test the --exclude-errors flag, which seems like a convenient way on its own to see if there are any undesirable cross-links between packages. I've recently split my e2e test environments into multiple frameworks, and making sure there isn't too much wild cross-pollination between them would be nice.

I had been using a linter config in some client projects for this purpose, can't recall which one though.

**Reply by reconsumeralization** [reply]
I really like the secret auto documentation the cursors using. I hear told, the pnpm isn't having dependencies issues with it's peer anymore.  😚🥰😍😂😘😚😁😁😁

**Comment by drwpow** [comment]
Devil’s advocate: why not enforce the boundaries by default? Make the “unsafe” behavior opt-in for users that are already using this, and make them aware of the drawbacks. As someone who’s dealt with this issue, there are no downsides to forcing that all packages either import local code, or import via a valid npm specifier (to all the points stated in the RFC).

[Vite has enforced this since v2.7](https://github.com/vitejs/vite/blob/v2/docs/config/index.md#serverfsstrict) (and by extension, Vitest), so a lot of people are already used to build systems that require this.

This may be a “breaking” or annoying change, but on the flipside, enabling by default could just improve the perception that people have of Turbopack’s quality—it guides you into safer monorepo strategies that are easier to optimize and scale. Or people will just get annoyed; hard to say 😅

**Reply by anthonyshew** [reply]
While it does sound attractive to make the recommended Boundaries a default out of the box, this feature might be unrealistic for us to ship as the default right away. Our goal is that folks should be able to get started with Turborepo with as little friction as possible, especially in huge repos (where Turborepo can have tremendous impact).

The problem ends up being that large repos can have the scariest, darkest corners. I've seen...things. 😄 That ends up meaning a feature like this being on by default is a barrier to entry and we would have to balance that against the cache safety reasoning that you've adeptly mentioned. The tradeoff for raising the quality bar of the average Turborepo would be that difficulty of adoption increases.

Right now, that balance is an unknown. This feature is only an idea, and we think we can reasonably expect how folks are going to use the feature - but we can only make an educated guess based on our experiences. Instead of going for making this the default straight out of the box, we can incrementally ship this such that folks who want it can turn it on at their leisure.

One way to think about this is in the cost of "getting it wrong". There are two scenarios:
1. We be cautious, and incrementally ship the recommended Boundaries through an opt-in flag. Folks can incrementally adopt Boundaries (or not) when we ship the feature for the first time. It doesn't require a major release to get the feature into your hands. After some time, we may find that this really should be the default and ship a major that includes the default flipping. Note that that is one, optional major release.
2. We be aggressive and ship the recommended boundaries on Day One. That makes one major release. It's possible that we find out that our instincts weren't correct that this should be enabled by default and it requires another major version of Turborepo to undo our mistake. That means, at a minimum, one forced major, plus a possibility for a second.

Hopefully that paints a picture of why I'm expecting we would eventually, slowly release this feature (if this RFC makes it to implementing at all, of course). I do agree with your point that cache safety is vital, and I wish it were as easy as making sure every repo gets it right at all times, but, today, I do believe we'll need to be thoughtful about how we approach the release of this feature.

**Comment by Netail** [comment]
It does seem like the boundaries options are missing from the schema.json,

```json
{
    "$schema": "https://turborepo.org/schema.json",
	// Property boundaries is not allowed.
    "boundaries": true
}
```

**Reply by anthonyshew** [reply]
This is only a request for comments at the moment! Not implemented.

**Reply by Netail** [reply]
Ahh okay, sorry. Got a bit confused if this was a thing already, tried it, but noticed the option not being available in the schema

**Reply by UnknownHero** [reply]
@anthonyshew Could you please provide the schema after the experimental implementation? Additionally, that schema will be needed for the documentation.

**Reply by trevorr** [reply]
The [public docs](https://turbo.build/repo/docs/reference/boundaries) now include using `"boundaries": { ... }` for tag rules. Will providing an object imply `true`, or is there an inconsistency here that needs to be addressed?

**Reply by anthonyshew** [reply]
The documentation is now the source of truth.

**Comment by vitalets** [comment]
For a single-package scenario, I recommend checking out [eslint-plugin-boundaries](https://github.com/javierbrea/eslint-plugin-boundaries). Its README is very descriptive and covers many interesting cases.

**Comment by anthonyshew** [comment]
All, we've updated the RFC and documentation to reflect the current state of the feature.

As of Turborepo 2.4.2, you can now try out built-in and custom Boundaries! Please try them out and give us your feedback.

**Reply by Netail** [reply]
We currently use `@<directory>` as an alias in our tsconfig to map to another directory, but it seems like turbo thinks these are actual packages.

e.g.
```json
// TSConfig
{
    "compilerOptions": {
        "paths": {
            "@components/*": [
                "./src/components/*"
            ],
		}
	}
}
```

```
cannot import package `@components/a-nice-compoment` because it is not a dependency
```

**Comment by ivancuric** [comment]
How would one go about ignoring these kinds of meta-imports?
 
```sh
  × cannot import package `uno.css` because it is not a dependency
    ╭─[/path/to/file.tsx:16:1]
 15 │ // this triggers extraction of CSS from the UnoCSS plugin
 16 │ import "uno.css";
    · ────────┬────────
    ·         ╰── package imported here
 17 │ 
    ╰────
```

**Reply by ebg1223** [reply]
50 | import { randomUUIDv7 } from "bun";
    : ^^^^^^^^^^^^^^^^^|^^^^^^^^^^^^^^^^^
    :                  `-- package imported here

Another use-case for ignoring certain imports!

**Comment by thebjorn** [comment]
Just tried it out and fixed two bugs immediately.

The two issues I found were import of json:
```
error: Expected ';', got 'with'
  --> E:\tilly\tools\nt\index.js:26:34
   |
26 | import pkg from './package.json' with {type: 'json'}
```
and it doesn't want to play with svelte files :-(

**Reply by anthonyshew** [reply]
Thanks for finding this! Should be fixed in https://github.com/vercel/turborepo/pull/10078.

**Comment by thebjorn** [comment]
Another issue... you're really stretching the usefulness of .json as a file format for programming dependency rules. Perhaps it's time to switch to a format that has the possibility for comments (so future devs can figure out the syntax without an accompanying `turbo.json.README` file...?)

**Reply by anthonyshew** [reply]
There's [a .jsonc request back here](https://github.com/vercel/turborepo/issues/3793) that we could make time for if it garners enough interest - or would make for a great community contribution!

**Comment by maxfriedmann** [comment]
Coming from Nx this seems to be the last missing piece!!!

I get an error on ubuntu, locally(windows) everything is fine . 

```
Run turbo boundaries
  x cannot import package `dns` because it is not a dependency
   ,-[~/apps/some-app/src/lib/something.server.ts:5:1]
 4 | (...)
 5 | import { Resolver } from "dns";
   : ^^^^^^^^^^^^^^^|^^^^^^^^^^^^^^^
   :                `-- package imported here
 6 | 
   `----

  x cannot import package `fs` because it is not a dependency
   ,-[~/apps/some-app/vite.config.ts:5:1]
 4 | import tailwindcss from "@tailwindcss/vite";
 5 | import { readFileSync } from "fs";
   : ^^^^^^^^^^^^^^^^^|^^^^^^^^^^^^^^^^
   :                  `-- package imported here
 6 | import { fileURLToPath } from "url";
   `----

  x cannot import package `url` because it is not a dependency
   ,-[~/apps/some-app/vite.config.ts:6:1]
 5 | import { readFileSync } from "fs";
 6 | import { fileURLToPath } from "url";
   : ^^^^^^^^^^^^^^^^^^|^^^^^^^^^^^^^^^^^
   :                   `-- package imported here
 7 | import type { PluginOption } from "vite";
   `----
```

Has anyone ever got that?

**Reply by maxfriedmann** [reply]
Prefixed the imports with `node:`, that did it. I still wonder why this is an ubuntu thing. NPM and node versions are in sync

**Comment by tkloht** [comment]
Thanks for the RFC 🙌 I like the direction a lot but have a few things that I'm not sure about:

First problem is having to define tags in each package:
- Our repo has ~190 packages with a folder structure like `apps/* , shared/*` where subfolders of apps and shared are packages 
- we'd want to prevent imports between apps and from  shared to apps, for example
- to do that we'd have to add a turbo.json to each package and define tags in them
- I think creating all those turbo.json files is a bit inconvenient but in the end that's fine. What's more of a problem is that each package can deviate from the expected configuration, which would be to tag all packages in the apps folder as `app`, and all packages in shared as `shared`
- For us it would be much better to have a central configuration for the boundaries, for example by assigning tags to directory globs in the root turbo.json similar to a pnpm-workspaces definition. I'm not sure if this is the best solution though, mainly wanted to explain the problem.

Another problem for us is not being able to configure the rule to allow imports only if the dependency is declared in the nearest package.json:
- Generally this rule makes sense of course, but we have some packages creating circular dependency graphs and work around it by declaring one of them in the root to break the cycle
- We're aware that this is not ideal but it's a transition process(*), we have our own tooling to prevent adding new imports like these
- (*) we have a small number of packages where I doubt we'll ever be able to declare them in all packages, these are mostly utilities related to storybook which are imported throughout the repository. They do not create problems other than the circular dependencies breaking turbo
- We can add these to the allow-list, but any usage in the repo will create a new violation. It would be much better DX if we could exclude some packages from the rule so they can be imported everywhere but only be declared in the root package.json
- For our repo we already have our own tooling to make this work. It would help us if we could opt out of this particular rule and still be able to use the tags and relative-import rules.

Hope this is possible to understand, let me know if I can clarify anything 🙏

**Comment by altano** [comment]
Neat idea. When I used to work at Netflix it was a HUGE pain that most repos used `npm` and did not hook up `eslint-plugin-import`, so dependencies of published packages could just be wrong. It's nice to see someone else trying to tackle that problem.

I found a bug in my testing:
`turbo boundaries` errors when a package imports itself. This is a valid use-case and has helpful properties, e.g. it allows you to test the PUBLIC package interface from a unit test (e.g. it verifies that you export what you think you export). No package.json entry is required for self-referencing, but you do have to be using a valid export.

**Comment by simbo** [comment]
I have to admit, I am new into monorepos and I am not sure, if I'm doing it wrong or have wrong expectations. Please enlighten me. :)

I think, boundaries is a great and promising feature.
Nevertheless, I have some issues with it:

- I have config files in my packages that extend root configs.
  For example, `vitest.config.ts`:
  ```ts
  import { defineConfig } from 'vitest/config';
	
  // @boundaries-ignore root config inheritance
  import baseConfig from '../../vitest.config.js';

  export default defineConfig({
    ...baseConfig,
    test: {
      ...baseConfig.test,
      include: ['**/*.test.ts']
    }
  });
  ```

- In my tests, I import from `vitest`, which is a devDependency of the root package and not of workspace packages. I do not see anything wrong with that...

- I test modules by importing them via their public package name (like @altano described before).

Adding ignore comments everywhere is not an option for me.
Every test file would need at least two boundaries ignore comments.

I would like to see all three things being solved with some config options based on glob patterns.

For example:
- for `**/*.test.ts`:
  - allow importing module `vitest` although not in package.json
  - allow importing the package itself via public name
- for `*.config.(ts|js)`:
  - allow importing from relative paths leading outside of the package

I would _not_ prefer generic file/folder ignores.
Let's be as specific as possible.

Thanks for requesting comments!

**Reply by anthonyshew** [reply]
> I have config files in my packages that extend root configs.

> In my tests, I import from vitest, which is a devDependency of the root package and not of workspace packages. I do not see anything wrong with that...

These are examples of specific patterns that we are trying to discourage because they are anti-patterns in JavaScript Workspaces.

Root files and dependencies are globals that affect all packages in the workspace, making it too easy to produce poor structural patterns in the repository. For this reason, we want to help Turborepo users to move away from these patterns and towards more modern practices.

**Reply by altano** [reply]
And there are many reasons other than caching that make that bad, eg extracting a package out of a monorepo can become impossible, typechecking files outside of your source root is probably not possible, node module resolution is not made for it and will probably break in weird ways, etc. If your vitest file imports things, where do you expect those to resolve? Where would they be declared as dependencies? In the root package.json? Do you now have two package.json files driving one node process’ resolution? It’s just not a good idea.

**Comment by someone635** [comment]
I am also new to monorepos, but one thing I'd like from this feature is a layer field in addition to the tags field. A package on layer 2 can import packages on layer 1, but packages on layer 1 can't import packages on layer 2. It is possible to achieve this with tags "layer-0", "layer-1", etc. but then you have to list them all explicitely to allow/deny them. So if you have 20 layers, you have to list all layers below in allow and list all layers above in deny. Or maybe there is already a simpler way to do this that I am not seeing?

**Reply by anthonyshew** [reply]
Can you explain more about your use case for "layers"? What are you trying to accomplish?

**Reply by someone635** [reply]
I am trying to refactor a complex app into unit packages with turborepo and it's been a great experience. But now I have a lot of small tiny packages that depend on each other in a chaotic way. I want to place all the packages in a clear low-level to high-level hierarchy (layers), to ensure I don't get accidental circular dependencies in the future, and just to make all the pieces of my application easier to reason about. So I put low-level packages, like my db connection, or 3rd party package wrappers, in the first few layers, then my backend packages that depend on those on layers above, and then front-end packages that depend on the backend on even greater layers.  So instead of a packages/ folder with all my packages, I have divided the packages/ folder in multiple folders (package groups) named layer-0, layer-1, layer-2, etc. where no package in layer-0 can import packages from layer-1 and above. I also added the layer number to my package name, so that I can manually enforce the rule. If I'm in layer 2 and try to import a package from layer-3, I know there is something wrong in my architecture and I need to refactor some more.

Maybe this is an unconventional architecture, but for me it seems very clean and it has been very helpful to guide me in my transition to turborepo. If I had a package with elements that spanned multiple layers, I knew my package was doing too much and I could split it in tinier packages. With a layer field added to Boundaries, I could remove the layer-* folders and the layer number from my package name, and stop enforcing the layering rule manually, I could let Turborepo warn me about problems.

I could do it with tags, but right now I have about 20 layers, so it would be a bit cumbersome to say the least. Or maybe there is a way I just don't see?

Anyway, I have no idea if this would be a popular feature or not, but I'm proposing it because it doesn't feel like a difficult feature to implement on top of tags, and it could maybe help others organize their packages or transition their codebase to turborepo the way it helped me.

**Reply by someone635** [reply]
Ok, I sandboxed more in my repo and won't be needing this anymore. Leaving the proposition here in case anyone else has a use case for it.

**Comment by stefan-karlsson** [comment]
First of all, great initiative — I’ve been looking forward to this feature! ~~I’m currently testing it and encountered an issue when running ESLint via Turborepo:~~

**Edit:** My mistake — I missed the [relevant documentation](https://turborepo.com/docs/reference/configuration#extends). I resolved the issue by [adding the extends property](https://github.com/stefan-karlsson/aviene/commit/be8181ffe8678808df5f7985edec09c445d85a5e) to `turbo.json`.

![image](https://github.com/user-attachments/assets/7f0a5c0d-d861-4316-93cb-4e16a2bb7b77)

View [the state of my monorepo at the time of the error here](https://github.com/stefan-karlsson/aviene/tree/caf826ea8bcd39ab95921459315adb168dc28aa6).

:bulb: My first impression is that this functionality would be better suited at the workspace level, rather than requiring a separate `turbo.json` file in each package. I’d love to see support for configuring this in the root `turbo.json`, ideally with the ability to tag packages using path globs.

**Reply by joshlperry** [reply]
Even if we could add tags in each package.json would be nicer than another config file

**Comment by astanb** [comment]
Another vote here for defining in the root turbo.json and/or package level package.json. Having moved from NX to Turborepo I really want to avoid the bloat of having so many config files in a package, so having the only way of configuring this by using a turbo.json in each package is not ideal for us.

**Comment by Raanelom** [comment]
We're currently using turbo boundaries in our CI/CD-pipeline, and also want to use it for local development as a code-quality step for better guarding of package scopes. So far, so good!
Lately, we upgraded to Turbo (2.8.12). With this version (and also in canary), the turbo boundaries command also scans all my dist and output folders that are included in my .gitignore. Is it possible to exclude certain directories from the turbo boundaries command? It would for example make sense to ignore gitignored-files, imho. On a clean (mono)repo it runs fine, but after a single build all my output files are scanned as well.

**Reply by MishaVyb** [reply]
the same issue

**Reply by AndreaPontrandolfo** [reply]
same, can this be looked at?

**Comment by corma-dev-account** [comment]
Just coming here to propose a feature on this : `run` with `--filter` tag-based.

It's completely outside of "boundaries" scope; but [the original "tags / scopes" discussion](https://github.com/vercel/turborepo/discussions/7460) was closed.

Simple use-case :
Running a big monorepo with multiple executables inside; I was to "run everything related to the web app"; "run everything related to backend integration sync", etc. with some overlaps.

**Comment by Offirmo** [comment]
I tried it on my ~180 pkg private monorepo. I'd need a glob exclusion: I keep "draft" non-compliant code in my packages under a specific folder naming (`tosort/YYYY`) and it trips boundaries even while this code is not used.

Apart from my personal case, there are so many specific tools/configs that could trigger this that a package-relative include/exclude pattern would be great.

**Comment by taco3064** [comment]
The RFC asks for continued feedback, so here is a design note on what a *passing* Boundaries run should say.

In the public reports I can find, the visible failures skew heavily toward false positives — tsconfig path aliases flagged as undeclared dependencies, `bun` imports read as type-only, subpath imports, import attributes. That skew is structural rather than a quality signal: a false positive blocks someone, so it gets reported; a false negative looks like success, so surfacing it usually takes a deliberate audit.

[#13417](https://github.com/vercel/turborepo/issues/13417) is the case in point, and it's worth reading as more than the bug it was. A dynamic import breaking out of its package produced:

```text
Checked 2 files in 2 packages, no issues found
```

Changing only that import to a static one produced the expected violation. Same rule, same breach, two spellings — and the passing output doesn't merely stay silent; it reports a file count, which invites the reader to infer coverage. Note also how it surfaced: a purpose-built reproduction repository. Nobody stumbles into that; you build it after you already suspect. It has since been fixed, which is the point — fixing the instance doesn't change what a pass means the next time the model of the graph falls short of the graph.

The Boundaries reference documents the two violation types and tags, with no limitations section and no statement of what a clean run covers. That makes the command output carry the contract by itself — and `no issues found` reads as a claim about the dependency graph.

On the cost of that, from the other side: I maintain a boundary checker for Vue/React that carried a silent false negative for twenty published releases. Nuxt's auto-imports leave no `import` statement, so the graph came back near-empty and clean. Nobody reported it — a clean report doesn't generate bug reports. We found it by auditing, and the fix was to detect the framework and refuse to initialize rather than emit a check that sees nothing.

So the missing contract isn't another ignore option. It's for `turbo boundaries` to report what it inspected, what it skipped, and which import forms or generated areas fell outside its model. Otherwise `no issues found` reads as a statement about the whole dependency graph, when it may only describe the subset the checker could see.

(The checker is [`@kekkai/blueprint`](https://github.com/taco3064/blueprint), MIT — cited as the source of the lived cost above, not as a suggestion for Turborepo.)
